tidy3d/components/medium.py

```diff
@@ -1040,14 +1391,128 @@ def pole_residue(self):
+        \\frac{\\Delta\\epsilon_i f_i^2}{f_i^2 - 2jf\\delta_i - f^2}
+
+    Example
+    -------
+    >>> x = np.linspace(-1, 1, 5)
+    >>> y = np.linspace(-1, 1, 6)
+    >>> z = np.linspace(-1, 1, 7)
+    >>> eps_inf = SpatialDataArray(np.ones((5, 6, 7)), coords=dict(x=x, y=y, z=z))
+    >>> epsilon_i = SpatialDataArray(np.random.random((5, 6, 7)), coords=dict(x=x, y=y, z=z))
+    >>> f_i = SpatialDataArray(1+np.random.random((5, 6, 7)), coords=dict(x=x, y=y, z=z))
+    >>> delta_i = SpatialDataArray(np.random.random((5, 6, 7)), coords=dict(x=x, y=y, z=z))
+    >>> lorentz_medium = CustomLorentz(eps_inf=eps_inf, coeffs=[(epsilon_i,f_i,delta_i),])
+    >>> eps = lorentz_medium.eps_model(200e12)
+    """
+
+    eps_inf: SpatialDataArray = pd.Field(
+        ...,
+        title="Epsilon at Infinity",
+        description="Relative permittivity at infinite frequency (:math:`\\epsilon_\\infty`).",
+        units=PERMITTIVITY,
+    )
+
+    coeffs: Tuple[Tuple[SpatialDataArray, SpatialDataArray, SpatialDataArray], ...] = pd.Field(
+        ...,
+        title="Coefficients",
+        description="List of (:math:`\\Delta\\epsilon_i, f_i, \\delta_i`) values for model.",
+        units=(PERMITTIVITY, HERTZ, HERTZ),
+    )
+
+    @pd.validator("eps_inf", always=True)
+    def _eps_inf_positive(cls, val):
+        """eps_inf should be positive"""
+        if np.any(val < 0):
+            raise SetupError("'eps_inf' should be positive.")
+        return val
+
+    @pd.validator("coeffs", always=True)
+    def _coeffs_correct_shape_and_sign(cls, val, values):
+        """coeffs should have consistent shape and sign."""
+        if "eps_inf" not in values:
+            raise ValidationError("'eps_inf' failed validation.")
+
+        expected_shape = values["eps_inf"].shape
+        for (de, f, delta) in val:
+            if (
+                de.shape != expected_shape
+                or f.shape != expected_shape
+                or delta.shape != expected_shape
+            ):
+                raise SetupError(
+                    "All terms in 'coeffs' should have the same dimension; "
+                    "The dimension should also be consistent with 'eps_inf'."
+                )
+            if np.any(de < 0):
+                raise SetupError(":math:`\\Delta\\epsilon_i` should be non-negative.")
+            if np.any(delta < 0):
+                raise SetupError(":math:`\\delta_i` should be non-negative.")
+        return val
+
+    @pd.validator("coeffs", always=True)
+    def _coeffs_delta_smaller_than_fi(cls, val):
+        """We restrict fi>delta_i for now."""
+        for (_, f, delta) in val:
+            if np.any(delta > f):
+                raise SetupError("We restrict :math:`\\delta_i<f_i` for now.")
+        return val
+
+    def _pole_residue_dict(self) -> Dict:
+        """Dict representation of Medium as a pole-residue model."""
+
+        poles = []
+        for (de, f, delta) in self.coeffs:
+
+            w = 2 * np.pi * f
```

Comment:
is there any way to reuse this from the regular `Lorentz` class? I remember that the equation was a bit error-prone so it would be good to not have separate copies.

Same comment for the other dispersive subclasses

Reply:
For Lorentz dispersion, one complexity is that there will be 2 poles if delta>f, and 1 pole if delta<f. 

- For custom Lorentz dispersion, it means the number of poles can vary spatially. So I temporarily restrict that delta<f at all spatial points. What do you think here?
- In regular Lorentz model, we have the conditional `d>w` etc. to treat the two. When `d` and `w` are spatial array, should i change this condition in regular Lorentz model to be `np.all(d>w)` in order to reuse?

Reply:
In Drude model, the problem of reusing comes from https://github.com/flexcompute/tidy3d/blob/58590ae54a2628e1caaf12da7b0ead256bbe9c3e/tidy3d/components/medium.py#L1096 so every `a0` needs to be converted to a dataarray.

Reply:
can the issue in regular Drude be mitigated through
```py
if isinstance(c0, complex):
    a0 = 0j
else:
    a0 = 0 * c0.copy()
```
or is that making things too complex?

Reply:
For Lorenz, I think it seems a bit too complicated to try to handle both 2 pole and 1 pole cases varying spatially using data arrays. So perhaps for now we just enforce that either all delta < f or all delta > f? Unless there is a way to unify them (eg use two poles for all, but have the 2nd poles be amplitude=0 for delta < f?

Reply:
> can the issue in regular Drude be mitigated through
> 
> ```python
> if isinstance(c0, complex):
>     a0 = 0j
> else:
>     a0 = 0 * c0.copy()
> ```
> 
> or is that making things too complex?

Sounds good!

Reply:
For Lorentz model, yeah, I prefer "either all delta < f or all delta > f" for now and extend later. A method is added:
```
if isinstance(coeff_a, SpatialDataArray):
    return np.all(coeff_a.data> coeff_b.data)
return coeff_a > coeff_b
```
`d>w` is then replaced by this method so that I can reuse from the regular `Lorentz` model.